### DISCUSSION TITLE
How to generate key using a passpharse as discussed in `Backup` Topic?

### DISCUSSION THREAD

**Discussion by KAGEYAM4** [missing-semester/missing-semester#248] [Q&A] [answered]
I have been looking for a program to generate a key from passpharse, like the ssh and password managers uses in the background.

In the topic, it was mentioned that it is a good idea to encrypt backup using key that is generated from masterpassword, so in the end we only have to remeber have master password. ( especially using some PBKDF and not just using hashing )

But i can't find any tools, most i found are generating public-private-key pair which itself uses passphrase to encrypt the private part of public-private key pair.

**Comment by anishathalye** [comment]
What is your goal? If you want to encrypt a file with a passphrase, there are tools out there that support this functionality directly, e.g., https://github.com/FiloSottile/age.

**Reply by KAGEYAM4** [reply]
Sorry for late reply.

>a useful trick is to protect the drives or accounts where the database is backed up using a simple cryptographic hash of the master password. Maintain a list of all the backup locations: if one day you fear that the master passphrase has been compromised you will have to change it immediately on all the database backups and the locations protected with keys derived from the master password.
- [Source](https://wiki.archlinux.org/title/security#Maintaining_passwords)

So, i want to generate a hash of master password and then i remebered about PBKDF2.

I tried finding programs which can generate strong key like what ssh-keygen, bitwarden and other password manager uses as a current standard for generating strong keys from master password.

**Reply by KAGEYAM4** [reply]
My last reply was a mess. I didn't worded my question correctly.

> PBKDF2 SHA-256 is used to derive the encryption key from your master password, however you may choose Argon2 as an alternative. Bitwarden salts and hashes your master password with your email address locally, before transmission to our servers.

Are there any programs that can take plaintext as input and  use `PBKDF2 SHA-256` or `Argon2` to generate cryptographic hash?

**Reply by KAGEYAM4** [reply, marked as answer]
Found it https://github.com/P-H-C/phc-winner-argon2 and corresponding arch-package https://archlinux.org/packages/core/x86_64/argon2/ & https://archlinux.org/packages/core/x86_64/nettle/
